packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {classNames, filterDOMProps, SlotProvider, useDOMRef, useSlotProps, useStyleProps} from '@react-spectrum/utils';
+import {DOMRef} from '@react-types/shared';
+import React, {useCallback, useEffect, useState} from 'react';
+import {SpectrumButtonGroupProps} from '@react-types/buttongroup';
+import styles from '@adobe/spectrum-css-temp/components/buttongroup/vars.css';
+import {useProvider, useProviderProps} from '@react-spectrum/provider';
+
+let lastScale;
+
+function ButtonGroup(props: SpectrumButtonGroupProps, ref: DOMRef<HTMLDivElement>) {
+  let {scale} = useProvider();
+  props = useProviderProps(props);
+  props = useSlotProps(props, 'buttonGroup');
+
+  let {
+    children,
+    orientation = 'horizontal',
+    isDisabled,
+    ...otherProps
+  } = props;
+
+  let {styleProps} = useStyleProps(otherProps);
+  let domRef = useDOMRef(ref);
+  let [hasOverflow, setHasOverflow] = useState(false);
+
+  let checkForOverflow = useCallback(() => {
+    if (domRef.current && orientation === 'horizontal') {
+      setHasOverflow(false);
+      let buttonGroupChildren = Array.from(domRef.current.children);
+      let childrenY = buttonGroupChildren.map(child => child.getBoundingClientRect().top);
+      // If any button's top is different from the others, overflow is happening
+      if (!childrenY.every(itemY => itemY === childrenY[0])) {
+        setHasOverflow(true);
+      }
+    }
+  }, [domRef, orientation]);
+
+  useEffect(() => {
+    if (orientation !== 'vertical') {
+      if (scale !== lastScale) {
+        if (lastScale === 'large' && scale === 'medium') {
+          setHasOverflow(false);
+        } else {
+          checkForOverflow();
+        }
+      }
+      lastScale = scale;
+    }
+  }, [setHasOverflow, checkForOverflow, scale, orientation]);
```

Comment:
I think this will have some trouble if children are added/removed or if the locale changes or if any strings change

example where that might happen, confirm button in dialog is pressed, it performs an async action before closing, and while it does that, it displays a spinner and says wait while disabled. If the buttons were horizontal and the wait made it wrap, then we'd end up in a situation that doesn't look very good

Reply:
We may need to use a MutationObserver https://caniuse.com/#feat=mutationobserver because the internal change in the confirm button to display a wait spinner may not trigger anything here

Reply:
When a change occurs, we'll need to remove the vertical class, useLayoutEffect to see if there's overflow and quickly reapply the vertical class if there is overflow